cranelift/codegen/src/isa/aarch64/lower.isle

```diff
@@ -21,3 +21,141 @@
 
 (rule (lower (has_type ty (null)))
       (value_reg (imm ty 0)))
+
+;;;; Rules for `iadd` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; `i64` and smaller
+
+;; Base case, simply adding things in registers.
+(rule (lower (has_type (fits_in_64 ty) (iadd x y)))
+      (value_reg (alu_rrr (iadd_op ty) (put_in_reg x) (put_in_reg y))))
+
+;; Special cases for when one operand is an immediate that fits in 12 bits.
+(rule (lower (has_type (fits_in_64 ty) (iadd x (imm12_from_value y))))
+      (value_reg (alu_rr_imm12 (iadd_op ty) (put_in_reg x) y)))
+
+(rule (lower (has_type (fits_in_64 ty) (iadd (imm12_from_value x) y)))
+      (value_reg (alu_rr_imm12 (iadd_op ty) (put_in_reg y) x)))
+
+;; Same as the previous special cases, except we can switch the addition to a
+;; subtraction if the negated immediate fits in 12 bits.
+(rule (lower (has_type (fits_in_64 ty) (iadd x (imm12_from_negated_value y))))
+      (value_reg (alu_rr_imm12 (isub_op ty) (put_in_reg x) y)))
+
+(rule (lower (has_type (fits_in_64 ty) (iadd (imm12_from_negated_value x) y)))
+      (value_reg (alu_rr_imm12 (isub_op ty) (put_in_reg y) x)))
+
+;; Special cases for when we're adding an extended register where the extending
+;; operation can get folded into the add itself.
+(rule (lower (has_type (fits_in_64 ty) (iadd x (extended_value_from_value y))))
+      (value_reg (alu_rr_extend_reg (iadd_op ty) (put_in_reg x) y)))
+
+(rule (lower (has_type (fits_in_64 ty) (iadd (extended_value_from_value x) y)))
+      (value_reg (alu_rr_extend_reg (iadd_op ty) (put_in_reg y) x)))
+
+;; Special cases for when we're adding the shift of a different
+;; register by a constant amount and the shift can get folded into the add.
+(rule (lower (has_type (fits_in_64 ty)
+                       (iadd x (def_inst (ishl y (def_inst (iconst (lshl_from_imm64 <ty amt))))))))
+      (value_reg (alu_rrr_shift (iadd_op ty) (put_in_reg x) (put_in_reg y) amt)))
+
+(rule (lower (has_type (fits_in_64 ty)
+                       (iadd (def_inst (ishl x (def_inst (iconst (lshl_from_imm64 <ty amt))))) y)))
+      (value_reg (alu_rrr_shift (iadd_op ty) (put_in_reg y) (put_in_reg x) amt)))
+
+;; Fold an `iadd` and `imul` combination into a `madd` instruction.
+(rule (lower (has_type (fits_in_64 ty) (iadd x (def_inst (imul y z)))))
+      (value_reg (alu_rrrr (madd_op ty) (put_in_reg y) (put_in_reg z) (put_in_reg x))))
+
+(rule (lower (has_type (fits_in_64 ty) (iadd (def_inst (imul x y)) z)))
+      (value_reg (alu_rrrr (madd_op ty) (put_in_reg x) (put_in_reg y) (put_in_reg z))))
+
+;; Helper to use either a 32 or 64-bit add depending on the input type.
+(decl iadd_op (Type) ALUOp)
+(rule (iadd_op (fits_in_32 _ty)) (ALUOp.Add32))
+(rule (iadd_op _ty) (ALUOp.Add64))
```

Comment:
Sorry for not catching this earlier -- I just noticed this while reading through the ISLE while having the language semantics in mind: this combination of rules depends on the more-specific-before-less behavior, and will give incorrect codegen if the latter rule is used when the former could apply, no?

The intention in the language design at least was that the order of rule selection is a heuristic that *should* go more-specific-first, for better perf/isel results, but doesn't *have to*. Specifically (i) it's somewhat brittle to depend on a subtle heuristic when patterns only have a partial ordering (here one of them is clearly more specific than the other, but it's possible to write two patterns with "unordered specific-ness" wrt each other), and (ii) we have some testing ideas related to randomly permuting the rewrite path we take to test if the generated code is still valid.

Another way of seeing the problem is that when we get to verifying lowerings formally, the rule here for the 64-bit case basically says "`iadd_op` for any type can always give `Add64`". Each rule standing on its own should have verifiable semantics; we shouldn't have to implicitly depend on this rule not firing in some cases because of other rules.

The priority mechanism on the other hand *is* solid load-bearing core-language-semantics stuff, so I think one solution here is to use explicit priorities. I'd actually prefer a slightly different solution though, which is to have another extractor `(larger_than_32 ty)`; that makes the intent super-clear. Thoughts?

Reply:
(I also recognize this is a bit of a footgun that didn't come with sharp warnings, so apologies for that! In a future where we have formal verification of lowerings, or where we have rule-order fuzzing, this would've been caught fairly quickly at least...)